src/raftstore/coprocessor/split_check/table.rs

```diff
@@ -0,0 +1,480 @@
+            Some(current_encoded_key)
+        } else {
+            None
+        }
+    } else if is_table_key(current_encoded_key) {
+        // Now we meet the very first table key of this region.
+        Some(current_encoded_key)
+    } else {
+        None
+    };
+    if let Some(key) = split_key {
+        to_encoded_table_prefix(key).map(|k| keys::data_key(&k))
+    } else {
+        None
+    }
+}
+
+fn last_key_of_region(db: &DB, region: &Region) -> Result<Option<Vec<u8>>> {
+    let end_key = keys::enc_end_key(region);
+    let mut last_key = None;
+
+    let iter_opt = IterOption::new(Some(end_key), false);
+    let mut iter = box_try!(db.new_iterator_cf(CF_WRITE, iter_opt));
+
+    // the last key
+    if iter.seek(SeekKey::End) {
+        let key = iter.key().to_vec();
+        last_key = Some(key);
+    } // else { No data in this CF }
+
+    match last_key {
+        Some(lk) => Ok(Some(lk)),
+        None => Ok(None),
+    }
+}
+
+fn to_encoded_table_prefix(encoded_key: &[u8]) -> Option<Vec<u8>> {
+    if let Ok(raw_key) = Key::from_encoded(encoded_key.to_vec()).raw() {
+        table_codec::extract_table_prefix(&raw_key)
+            .map(|k| Key::from_raw(k).encoded().to_vec())
+            .ok()
+    } else {
+        None
+    }
+}
+
+// Encode a key like `t{i64}` will append some unecessary bytes to the output,
+// This function extracts the first 10 bytes, they are enough to find out which
+// table this key belongs to.
+const ENCODED_TABLE_TABLE_PREFIX: usize = table_codec::TABLE_PREFIX_KEY_LEN + 1;
+
+fn is_table_key(encoded_key: &[u8]) -> bool {
+    encoded_key.starts_with(table_codec::TABLE_PREFIX) &&
+        encoded_key.len() >= ENCODED_TABLE_TABLE_PREFIX
+}
+
+fn is_same_table(left_key: &[u8], right_key: &[u8]) -> bool {
+    is_table_key(left_key) && is_table_key(right_key) &&
+        left_key[..ENCODED_TABLE_TABLE_PREFIX] == right_key[..ENCODED_TABLE_TABLE_PREFIX]
+}
+
+#[cfg(test)]
+mod test {
+    use std::io::Write;
+    use std::sync::Arc;
+    use std::sync::mpsc;
+
+    use tempdir::TempDir;
+    use rocksdb::Writable;
+    use kvproto::metapb::Peer;
+
+    use storage::ALL_CFS;
+    use storage::types::Key;
+    use raftstore::store::{Msg, SplitCheckRunner, SplitCheckTask};
+    use util::rocksdb::new_engine;
+    use util::worker::Runnable;
+    use util::transport::RetryableSendCh;
+    use util::config::ReadableSize;
+    use util::codec::number::NumberEncoder;
+    use coprocessor::codec::table::{TABLE_PREFIX, TABLE_PREFIX_KEY_LEN};
+
+    use raftstore::coprocessor::{Config, CoprocessorHost};
+    use super::*;
+
+    /// Composes table record and index prefix: `t[table_id]`.
+    // Port from TiDB
+    fn gen_table_prefix(table_id: i64) -> Vec<u8> {
+        let mut buf = Vec::with_capacity(TABLE_PREFIX_KEY_LEN);
+        buf.write_all(TABLE_PREFIX).unwrap();
+        buf.encode_i64(table_id).unwrap();
+        buf
+    }
+
+    #[test]
+    fn test_last_key_of_region() {
```

Comment:
This test is too complicated. Actually `last_key_or_region` has nothing to do with tables, so you can just test if it works when there is no keys, one key or multiple keys in the region.